typed-racket-lib/typed-racket/base-env/unit-prims.rkt

```diff
@@ -0,0 +1,767 @@
+    [(_ name:id case ...)
+     #`(define-syntax (name stx)
+         (syntax-parse stx
+           [(_) #'(begin)]
+           [(name e)
+            (define exp-e 
+              (local-expand #'e (syntax-local-context) (kernel-form-identifier-list)))
+            (syntax-parse exp-e
+              #:literal-sets (kernel-literals)
+              [(~var b (begin-form #'name))
+               #'b.trampoline-form]
+              case ...
+              [_ 
+               exp-e])]
+           [(~var e (name-form #'name))
+            #'e.trampoline-form]))]
+    [(_ (name:id arg ...) case ...)
+     #`(define-syntax (name stx)
+         (syntax-parse stx
+           [(_ arg ...) #'(begin)]
+           [(name arg ... e) 
+            (define exp-e 
+              (local-expand #'e (syntax-local-context) (kernel-form-identifier-list)))
+            (syntax-parse exp-e
+              #:literal-sets (kernel-literals)
+              [(~var b (begin-form #'name #`(arg ...)))
+               #'b.trampoline-form]
+              case ...
+              [_ exp-e])]
+           [(_ arg ... (~var exprs (rest-form #'name #`(arg ...))))
+            #'exprs.trampoline-form]))]))
+
+
+;; Typed macro for define-values/invoke-unit
+;; This has to be handled specially because they types of
+;; the defined values must be registered in the environment
+(define-for-syntax (imports/members sig-id)
+  (match-define-values (_ (list imp-mem ...) _ _)
+                       (signature-members sig-id sig-id))
+  #`(#,sig-id #,@(map (lambda (id)
+                        (local-expand
+                         id
+                         (syntax-local-context)
+                         (kernel-form-identifier-list)))
+                      imp-mem)))
+
+(define-for-syntax (process-dv-exports es)
+  (for/list ([e (in-list es)])
+    (syntax-parse e
+      [s:sig-spec
+       (define sig-id #'s.sig-name)
+       (define renamer (attribute s.rename))
+       (match-define-values (_ (list ex-mem ...) _ _)
+                            (signature-members sig-id sig-id))
+       #`(#,sig-id #,@(map renamer ex-mem))])))
+
+(define-syntax (define-values/invoke-unit stx)
+  (syntax-parse stx 
+    #:literals (import export)
+    [(_ unit-expr
+        (import isig:sig-spec ...)
+        (export esig:sig-spec ...))
+     (define imports-stx (syntax->list #'(isig.sig-name ...)))
+     (define exports-stx (syntax->list #'(esig ...)))
+     (with-syntax ([temp (syntax-local-introduce (generate-temporary))])
+       #`(begin
+           #,(internal (quasisyntax/loc stx
+                         (define-values/invoke-unit-internal
+                           (#,@(map imports/members imports-stx))
+                           (#,@(process-dv-exports exports-stx)))))
+           (: temp (Unit (import isig.sig-name ...)
+                         (export esig.sig-name ...)
+                         ;; FIXME this needs to be AnyValues
+                         ;; FIXME: init-depends
```

Comment:
Seems like we should implement the `init-depends` support here. Is there anything blocking that? (the other FIXME makes  sense)